<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (1.8.0_192) on Thu Aug 04 13:00:37 GMT-06:00 2022 -->
<title>Dataset</title>
<meta name="date" content="2022-08-04">
<link rel="stylesheet" type="text/css" href="../../../../stylesheet.css" title="Style">
<script type="text/javascript" src="../../../../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="Dataset";
        }
    }
    catch(err) {
    }
//-->
var methods = {"i0":10,"i1":10,"i2":10,"i3":10,"i4":10,"i5":10,"i6":10,"i7":10,"i8":10,"i9":10,"i10":10,"i11":9,"i12":10,"i13":10};
var tabs = {65535:["t0","All Methods"],1:["t1","Static Methods"],2:["t2","Instance Methods"],8:["t4","Concrete Methods"]};
var altColor = "altColor";
var rowColor = "rowColor";
var tableTab = "tableTab";
var activeTableTab = "activeTableTab";
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/Dataset.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../index-all.html">Index</a></li>
<li><a href="../../../../help-doc.html">Help</a></li>
</ul>
<div class="aboutLanguage">BioLegato sequence canvas support v1.1.0 API</div>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../../../../org/biolegato/sequence/data/DataFormat.html" title="class in org.biolegato.sequence.data"><span class="typeNameLink">Prev&nbsp;Class</span></a></li>
<li><a href="../../../../org/biolegato/sequence/data/FastAFile.html" title="class in org.biolegato.sequence.data"><span class="typeNameLink">Next&nbsp;Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../../../../index.html?org/biolegato/sequence/data/Dataset.html" target="_top">Frames</a></li>
<li><a href="Dataset.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#fields.inherited.from.class.javax.swing.AbstractListModel">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="subTitle">org.biolegato.sequence.data</div>
<h2 title="Class Dataset" class="title">Class Dataset</h2>
</div>
<div class="contentContainer">
<ul class="inheritance">
<li>java.lang.Object</li>
<li>
<ul class="inheritance">
<li>javax.swing.AbstractListModel</li>
<li>
<ul class="inheritance">
<li>org.biolegato.sequence.data.Dataset</li>
</ul>
</li>
</ul>
</li>
</ul>
<div class="description">
<ul class="blockList">
<li class="blockList">
<dl>
<dt>All Implemented Interfaces:</dt>
<dd>java.io.Serializable, javax.swing.ListModel</dd>
</dl>
<hr>
<br>
<pre>public class <span class="typeNameLabel">Dataset</span>
extends javax.swing.AbstractListModel</pre>
<div class="block"><p>The internal document format for BioLegato.</p>

 <p>This document is structured as a linked list of sequences.  Each character
 has an offset based on its position within the list and it's position within
 its containing sequence.  Sequences start at 0 (first character in the first
 sequence in the list, and ens with the last character in the last sequence
 within the list.</p></div>
<dl>
<dt><span class="simpleTagLabel">Author:</span></dt>
<dd>Graham Alvare, Brian Fristensky</dd>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../../serialized-form.html#org.biolegato.sequence.data.Dataset">Serialized Form</a></dd>
</dl>
</li>
</ul>
</div>
<div class="summary">
<ul class="blockList">
<li class="blockList">
<!-- =========== FIELD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="field.summary">
<!--   -->
</a>
<h3>Field Summary</h3>
<ul class="blockList">
<li class="blockList"><a name="fields.inherited.from.class.javax.swing.AbstractListModel">
<!--   -->
</a>
<h3>Fields inherited from class&nbsp;javax.swing.AbstractListModel</h3>
<code>listenerList</code></li>
</ul>
</li>
</ul>
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor.summary">
<!--   -->
</a>
<h3>Constructor Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Constructor Summary table, listing constructors, and an explanation">
<caption><span>Constructors</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../org/biolegato/sequence/data/Dataset.html#Dataset-org.biolegato.sequence.canvas.SequenceCanvas-">Dataset</a></span>(<a href="../../../../org/biolegato/sequence/canvas/SequenceCanvas.html" title="class in org.biolegato.sequence.canvas">SequenceCanvas</a>&nbsp;canvas)</code>
<div class="block">Creates a new instance of the Dataset object.</div>
</td>
</tr>
</table>
</li>
</ul>
<!-- ========== METHOD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="method.summary">
<!--   -->
</a>
<h3>Method Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Method Summary table, listing methods, and an explanation">
<caption><span id="t0" class="activeTableTab"><span>All Methods</span><span class="tabEnd">&nbsp;</span></span><span id="t1" class="tableTab"><span><a href="javascript:show(1);">Static Methods</a></span><span class="tabEnd">&nbsp;</span></span><span id="t2" class="tableTab"><span><a href="javascript:show(2);">Instance Methods</a></span><span class="tabEnd">&nbsp;</span></span><span id="t4" class="tableTab"><span><a href="javascript:show(8);">Concrete Methods</a></span><span class="tabEnd">&nbsp;</span></span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tr id="i0" class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../org/biolegato/sequence/data/Dataset.html#add-int-org.biolegato.sequence.data.Seq-">add</a></span>(int&nbsp;y,
   <a href="../../../../org/biolegato/sequence/data/Seq.html" title="class in org.biolegato.sequence.data">Seq</a>&nbsp;seq)</code>
<div class="block">Adds a sequence to the Dataset sequence object container.</div>
</td>
</tr>
<tr id="i1" class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../org/biolegato/sequence/data/Dataset.html#addSequences-int-java.util.Collection-">addSequences</a></span>(int&nbsp;y,
            java.util.Collection&lt;<a href="../../../../org/biolegato/sequence/data/Seq.html" title="class in org.biolegato.sequence.data">Seq</a>&gt;&nbsp;seqs)</code>
<div class="block">Adds sequences to the Dataset sequence object container.</div>
</td>
</tr>
<tr id="i2" class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../org/biolegato/sequence/data/Dataset.html#delete-int-int-int-boolean-">delete</a></span>(int&nbsp;x,
      int&nbsp;y,
      int&nbsp;length,
      boolean&nbsp;protect)</code>
<div class="block">Deletes a portion of a sequence from the Dataset sequence object
 container.</div>
</td>
</tr>
<tr id="i3" class="rowColor">
<td class="colFirst"><code>java.lang.Object</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../org/biolegato/sequence/data/Dataset.html#getElementAt-int-">getElementAt</a></span>(int&nbsp;index)</code>
<div class="block">Returns the names of sequences based on their Y-coordinate/"line number".</div>
</td>
</tr>
<tr id="i4" class="altColor">
<td class="colFirst"><code>int[]</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../org/biolegato/sequence/data/Dataset.html#getgroup-int-">getgroup</a></span>(int&nbsp;lineNumber)</code>
<div class="block">Obtains all of the line numbers of the sequences in a particular group.</div>
</td>
</tr>
<tr id="i5" class="rowColor">
<td class="colFirst"><code><a href="../../../../org/biolegato/sequence/canvas/ColourMask.html" title="class in org.biolegato.sequence.canvas">ColourMask</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../org/biolegato/sequence/data/Dataset.html#getMask-int-">getMask</a></span>(int&nbsp;number)</code>
<div class="block">Retrieves the colour mask object for a given sequence 'line' in the
 Dataset sequence object container.</div>
</td>
</tr>
<tr id="i6" class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../org/biolegato/sequence/data/Dataset.html#getSequence-char:A-int-int-int-">getSequence</a></span>(char[]&nbsp;array,
           int&nbsp;number,
           int&nbsp;offset,
           int&nbsp;length)</code>
<div class="block">Retrieves the text contained within a sequence object in the Dataset.</div>
</td>
</tr>
<tr id="i7" class="rowColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../org/biolegato/sequence/data/Dataset.html#getSequenceLength-int-">getSequenceLength</a></span>(int&nbsp;number)</code>
<div class="block">Retrieves the length (in characters) of a given sequence 'line' in the
 Dataset sequence object container.</div>
</td>
</tr>
<tr id="i8" class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../org/biolegato/sequence/data/Dataset.html#getSize--">getSize</a></span>()</code>
<div class="block">Returns a count of the total number of sequences (lines) in the Dataset
 sequence object container.</div>
</td>
</tr>
<tr id="i9" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../org/biolegato/sequence/data/Dataset.html#group-int:A-">group</a></span>(int[]&nbsp;sequences)</code>
<div class="block">Creates a new sequence-group comprised of the sequences specified by
 Y-coordinate/"line numbers" in the array parameter 'sequences'.</div>
</td>
</tr>
<tr id="i10" class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../org/biolegato/sequence/data/Dataset.html#insert-int-int-char:A-int-int-boolean-">insert</a></span>(int&nbsp;x,
      int&nbsp;y,
      char[]&nbsp;text,
      int&nbsp;offset,
      int&nbsp;length,
      boolean&nbsp;protect)</code>
<div class="block">Inserts text into the Dataset sequence object container.</div>
</td>
</tr>
<tr id="i11" class="rowColor">
<td class="colFirst"><code>static boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../org/biolegato/sequence/data/Dataset.html#isProtectionsOn-org.biolegato.sequence.data.Seq.Type-boolean-boolean-boolean-char:A-int-int-">isProtectionsOn</a></span>(<a href="../../../../org/biolegato/sequence/data/Seq.Type.html" title="enum in org.biolegato.sequence.data">Seq.Type</a>&nbsp;type,
               boolean&nbsp;protect_align,
               boolean&nbsp;protect_ambig,
               boolean&nbsp;protect_uambig,
               char[]&nbsp;test,
               int&nbsp;start,
               int&nbsp;end)</code>
<div class="block">Checks a string against all of a sequence's protection settings.</div>
</td>
</tr>
<tr id="i12" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../org/biolegato/sequence/data/Dataset.html#removeSequences-int:A-">removeSequences</a></span>(int[]&nbsp;indices)</code>
<div class="block">Removes multiple sequences from the Dataset sequence object container.</div>
</td>
</tr>
<tr id="i13" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../org/biolegato/sequence/data/Dataset.html#ungroup-int:A-">ungroup</a></span>(int[]&nbsp;sequences)</code>
<div class="block">Ungroups an array of sequences by iteratively calling the single-sequence
 ungroup function for each sequence number in the array.</div>
</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.javax.swing.AbstractListModel">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;javax.swing.AbstractListModel</h3>
<code>addListDataListener, fireContentsChanged, fireIntervalAdded, fireIntervalRemoved, getListDataListeners, getListeners, removeListDataListener</code></li>
</ul>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.java.lang.Object">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;java.lang.Object</h3>
<code>clone, equals, finalize, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</code></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="details">
<ul class="blockList">
<li class="blockList">
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor.detail">
<!--   -->
</a>
<h3>Constructor Detail</h3>
<a name="Dataset-org.biolegato.sequence.canvas.SequenceCanvas-">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>Dataset</h4>
<pre>public&nbsp;Dataset(<a href="../../../../org/biolegato/sequence/canvas/SequenceCanvas.html" title="class in org.biolegato.sequence.canvas">SequenceCanvas</a>&nbsp;canvas)</pre>
<div class="block">Creates a new instance of the Dataset object.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>canvas</code> - the parent sequence canvas to associate with the Dataset
                 object with.</dd>
</dl>
</li>
</ul>
</li>
</ul>
<!-- ============ METHOD DETAIL ========== -->
<ul class="blockList">
<li class="blockList"><a name="method.detail">
<!--   -->
</a>
<h3>Method Detail</h3>
<a name="add-int-org.biolegato.sequence.data.Seq-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>add</h4>
<pre>public&nbsp;boolean&nbsp;add(int&nbsp;y,
                   <a href="../../../../org/biolegato/sequence/data/Seq.html" title="class in org.biolegato.sequence.data">Seq</a>&nbsp;seq)</pre>
<div class="block">Adds a sequence to the Dataset sequence object container.  This function
 also calls calls all of the listeners which are listening to the Dataset
 object.  Mostly (ane even 'completely', at the time of this writing),
 listeners are used for displaying the data in the sequence list and the
 sequence text area.  Note that the add method inserts a sequence object,
 as a whole, into the Dataset, while the insert method inserts the text
 of a sequence object into the middle of the already present sequence
 object in the Dataset.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>y</code> - the line index number to insert the sequence object at.</dd>
<dd><code>seq</code> - the sequence object to insert.</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>true if the insertion was successful, otherwise false.</dd>
</dl>
</li>
</ul>
<a name="insert-int-int-char:A-int-int-boolean-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>insert</h4>
<pre>public&nbsp;boolean&nbsp;insert(int&nbsp;x,
                      int&nbsp;y,
                      char[]&nbsp;text,
                      int&nbsp;offset,
                      int&nbsp;length,
                      boolean&nbsp;protect)</pre>
<div class="block"><p>Inserts text into the Dataset sequence object container.  This
 function also calls calls all of the listeners which are listening to the
 Dataset object.  Mostly (ane even 'completely', at the time of this
 writing), listeners are used for displaying the data in the sequence list
 and the sequence text area.  Note that the add method inserts a sequence
 object, as a whole, into the Dataset, while the insert method inserts the
 text of a sequence object into the middle of the already present sequence
 object in the Dataset.</p>

 <p>This function also obeys all sequence permissions/protections (as long
 as the protect boolean is set to true).  In the case of text insertion,
 text may NOT be inserted if it contains a protected character.  For
 example, if the sequence is a protein sequence and the unambiguous
 character protections are set to true, then the text "BZAZ" may not be
 inserted because it contains one unambiguous character (A, for alanine);
 however, the text "BZZZ" may be inserted, as long as the ambiguous
 character protections are set to false.  See Dataset.isProtectionsOn for
 more information about character protections.</p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>x</code> - the X-coordinate to insert the sequence text.</dd>
<dd><code>y</code> - the Y-coordinate to insert the sequence text.</dd>
<dd><code>text</code> - the sequence text to insert.</dd>
<dd><code>offset</code> - the offset within the array to insert.</dd>
<dd><code>length</code> - the length in characters of data from the array to insert.</dd>
<dd><code>protect</code> - whether to test the protections of the sequence, already
                in the Dataset, before inserting the text.</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>true if the insertion was successful, otherwise false.</dd>
</dl>
</li>
</ul>
<a name="delete-int-int-int-boolean-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>delete</h4>
<pre>public&nbsp;boolean&nbsp;delete(int&nbsp;x,
                      int&nbsp;y,
                      int&nbsp;length,
                      boolean&nbsp;protect)</pre>
<div class="block"><p>Deletes a portion of a sequence from the Dataset sequence object
 container.  This function also calls calls all of the listeners which are
 listening to the Dataset object.  Mostly (ane even 'completely', at the
 time of this writing), listeners are used for displaying the data in the
 sequence list and the sequence text area.  Note that the add method
 deletes a sequence object, as a whole, into the Dataset, while the delete
 method deletes the text of a sequence object into the middle of the
 already present sequence object in the Dataset.</p>

 <p>This function also obeys all sequence permissions/protections (as long
 as the protect boolean is set to true).  In the case of text deletion,
 text may NOT be deleted if it contains a protected character.  For
 example, if the sequence is a protein sequence and the unambiguous
 character protections are set to true, then the text "BZAZ" may not be
 deleted because it contains one unambiguous character (A, for alanine);
 however, the text "BZZZ" may be deleted, as long as the ambiguous
 character protections are set to false.</p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>x</code> - the X-coordinate to begin the the sequence text deletion.</dd>
<dd><code>y</code> - the Y-coordinate to begin the the sequence text deletion.</dd>
<dd><code>length</code> - is the length (number of characters) of text to delete.</dd>
<dd><code>protect</code> - whether to test the protections of the sequence, already
                in the Dataset, before deleting the text.</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>true if the deletion was successful, otherwise false.</dd>
</dl>
</li>
</ul>
<a name="addSequences-int-java.util.Collection-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>addSequences</h4>
<pre>public&nbsp;boolean&nbsp;addSequences(int&nbsp;y,
                            java.util.Collection&lt;<a href="../../../../org/biolegato/sequence/data/Seq.html" title="class in org.biolegato.sequence.data">Seq</a>&gt;&nbsp;seqs)</pre>
<div class="block">Adds sequences to the Dataset sequence object container.  This function
 also calls calls all of the listeners which are listening to the Dataset
 object.  Mostly (ane even 'completely', at the time of this writing),
 listeners are used for displaying the data in the sequence list and the
 sequence text area.  Note that the add method inserts a sequence object,
 as a whole, into the Dataset, while the insert method inserts the text
 of a sequence object into the middle of the already present sequence
 object in the Dataset.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>y</code> - the line index number to insert the sequences object at.</dd>
<dd><code>seqs</code> - the collection of sequence objects to insert.</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>true if the insertion was successful, otherwise false.</dd>
</dl>
</li>
</ul>
<a name="removeSequences-int:A-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>removeSequences</h4>
<pre>public&nbsp;void&nbsp;removeSequences(int[]&nbsp;indices)</pre>
<div class="block">Removes multiple sequences from the Dataset sequence object container.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>indices</code> - an array of line numbers to remove.</dd>
</dl>
</li>
</ul>
<a name="getSize--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getSize</h4>
<pre>public&nbsp;int&nbsp;getSize()</pre>
<div class="block">Returns a count of the total number of sequences (lines) in the Dataset
 sequence object container.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the total number of sequences in the Dataset object.</dd>
</dl>
</li>
</ul>
<a name="getSequenceLength-int-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getSequenceLength</h4>
<pre>public&nbsp;int&nbsp;getSequenceLength(int&nbsp;number)</pre>
<div class="block">Retrieves the length (in characters) of a given sequence 'line' in the
 Dataset sequence object container.  The sequence is specified by its
 Y-coordinate.  If the Y-coordinate value is invalid, a length of -1 is
 returned.  Invalid Y-coordinates are values which exceed the total number
 of sequences in the Dataset, or values less than zero.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>number</code> - the y coordinate to retrieve the sequence.</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the length of sequence (-1 if 'number' is invalid).</dd>
</dl>
</li>
</ul>
<a name="getMask-int-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getMask</h4>
<pre>public&nbsp;<a href="../../../../org/biolegato/sequence/canvas/ColourMask.html" title="class in org.biolegato.sequence.canvas">ColourMask</a>&nbsp;getMask(int&nbsp;number)</pre>
<div class="block">Retrieves the colour mask object for a given sequence 'line' in the
 Dataset sequence object container.  The sequence is specified by its
 Y-coordinate.  If the Y-coordinate value is invalid, the DEFAULT_MASK
 object is returned instead.  Invalid Y-coordinates are values which
 exceed the total number of sequences in the Dataset, or negative values.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>number</code> - the line number to retrieve the colour mask for.</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the colour mask for the given sequence.</dd>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../../org/biolegato/sequence/canvas/SequenceCanvas.html#DEFAULT_MASK"><code>SequenceCanvas.DEFAULT_MASK</code></a></dd>
</dl>
</li>
</ul>
<a name="getSequence-char:A-int-int-int-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getSequence</h4>
<pre>public&nbsp;int&nbsp;getSequence(char[]&nbsp;array,
                       int&nbsp;number,
                       int&nbsp;offset,
                       int&nbsp;length)</pre>
<div class="block"><p>Retrieves the text contained within a sequence object in the Dataset.
 The sequence object to extract text from is specified by its
 Y-coordinate.  The offset and number of characters to extract from the
 sequence are specified by 'offset' and 'length'.  The destination for the
 text extracted is the character array 'array'.  If the Y-coordinate or
 offset values are invalid, no data is extracted.  Invalid Y-coordinates
 are values which exceed the total number of sequences in the Dataset, or
 values less than zero.  Invalid offsets are those which exceed the number
 of characters on the sequence "line", or offset values less than zero.
 </p>

 <p>In contrast, if the length value exceeds the number of characters in
 the sequence, the sequence's length is used instead.  Thus, the length
 parameter may be slightly invalid (it may exceed the sequence length, but
 NOT be less than zero).</p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>array</code> - the destination for the characters to be copied to.</dd>
<dd><code>number</code> - the sequence "line number"/Y-coordinate to extract data.</dd>
<dd><code>offset</code> - the offset within the sequence to begin copying data from.</dd>
<dd><code>length</code> - the number of characters to attempt to copy.</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the number of characters actually copied.</dd>
</dl>
</li>
</ul>
<a name="group-int:A-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>group</h4>
<pre>public&nbsp;void&nbsp;group(int[]&nbsp;sequences)</pre>
<div class="block">Creates a new sequence-group comprised of the sequences specified by
 Y-coordinate/"line numbers" in the array parameter 'sequences'.  Please
 note that this method first "ungroups" all of sequences passed to it
 before creating a new group.  This ensures that no sequence belongs to
 more than one group at a time.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>sequences</code> - the array of Y-coordinates corresponding to sequences
                   to include in the new group.</dd>
</dl>
</li>
</ul>
<a name="ungroup-int:A-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>ungroup</h4>
<pre>public&nbsp;void&nbsp;ungroup(int[]&nbsp;sequences)</pre>
<div class="block">Ungroups an array of sequences by iteratively calling the single-sequence
 ungroup function for each sequence number in the array.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>sequences</code> - the array of Y-coordinates/"line numbers", which
                  correspond to the sequences to ungroup.</dd>
</dl>
</li>
</ul>
<a name="getgroup-int-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getgroup</h4>
<pre>public&nbsp;int[]&nbsp;getgroup(int&nbsp;lineNumber)</pre>
<div class="block">Obtains all of the line numbers of the sequences in a particular group.
 The group is specified to be the same group as the sequence on the line
 number ('lineNumber').  Thus, another way to word this is "this function
 obtains all of the sequences which belong to the same group as the
 sequence contained at the Y-coordinate value 'lineNumber'".</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>lineNumber</code> - the Y-coordinate/"line number" of the sequence to
                    use as a reference for determining the group number.</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>an array containing all of the line numbers in the group.</dd>
</dl>
</li>
</ul>
<a name="getElementAt-int-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getElementAt</h4>
<pre>public&nbsp;java.lang.Object&nbsp;getElementAt(int&nbsp;index)</pre>
<div class="block">Returns the names of sequences based on their Y-coordinate/"line number".
 This function is used to interface with the JList object.  Please note
 that the Sequence objects returned will print their sequence name and
 group number (in a formatted string), when their toString methods are
 called.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>index</code> - the Y-coordinate of the sequence to obtain.</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>returns the sequence object at the specified Y-coordinate.
         In turn, this sequence object will return its sequence name when
         its toString method is called.</dd>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../../org/biolegato/sequence/data/Seq.html#toString--"><code>Seq.toString()</code></a></dd>
</dl>
</li>
</ul>
<a name="isProtectionsOn-org.biolegato.sequence.data.Seq.Type-boolean-boolean-boolean-char:A-int-int-">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>isProtectionsOn</h4>
<pre>public static&nbsp;boolean&nbsp;isProtectionsOn(<a href="../../../../org/biolegato/sequence/data/Seq.Type.html" title="enum in org.biolegato.sequence.data">Seq.Type</a>&nbsp;type,
                                      boolean&nbsp;protect_align,
                                      boolean&nbsp;protect_ambig,
                                      boolean&nbsp;protect_uambig,
                                      char[]&nbsp;test,
                                      int&nbsp;start,
                                      int&nbsp;end)</pre>
<div class="block"><p>Checks a string against all of a sequence's protection settings.  A
 string of text is said to violate the protection settings (for deletion
 or insertion purposes) if the text contains any characters which are
 protected.  Whether a character is protected is determined by the Seq
 object's protection status settings.  These protection status settings
 are split into three variables (for 3 classes of characters):</p>
 <ol>
      <li><b>protect_align</b> -- for alingment characters.  These
              are the characters which represent alignment gaps, such as
              dashes and whitespace.  For a complete list of alignment
              characters, please see the Seq object's protect_align
              field specification.</li>

      <li><b>protect_ambig</b> -- for ambiguous sequence characters.
              Ambiguous sequence characters are characters which can
              represent more than one item (such as nucleotide or amino
              acid) in a sequence.  For example, 'N' is an ambiguous DNA
              character, because it can represent any DNA nucleotide.
              For a complete list of ambiguous sequence characters, please
              see the Seq object's protect_ambig field specification.</li>

      <li><b>protect_uambig</b> -- for unambiguous sequence characters.
              Unambiguous sequence characters are characters which can
              represent only one item (e.g. nucleotide or amino acid) in a
              sequence.  For example, 'A' is an unambiguous DNA character,
              because it only can represent one DNA nucleotide (Adenine).
              For a complete list of unambiguous sequence characters,
              please see the Seq object's protect_unambig field
              specification.</li>
 </ol>
 <p>If a character from any of the three classes is matched in the text
 passed to this function, and the character matched has its protection
 status is true, the text is barred from insertion or deletion.  In
 contrast, any character matched in the text with its character protection
 status set to false is ignored.  Thus text containing all characters
 which have false protection statuses may be inserted or deleted.  A
 couple of examples will help demonstrate this:</p>
 <ol>
      <li>Consider adding the text 'BZZ' to a protein sequence.  If the
          ambiguous character protection is off (false), then the text
          should be added without problem, regardless of the protection
          status for alignment or unambiguous characters.  Thus, this
          function will return the value of false.</li>
      <li>Consider adding the text 'BZAZ' to a protein sequence.  If the
          ambiguous character protection is off (false), but the
          unambiguous character protection status is on (true) then the
          text cannot be added because it contains the amino acid Alanine,
          which is a protected unambiguous amino acid.  Thus the function
          will return the value of true.</li>
      <li>Now, consider adding the same text as above ('BZAZ') to the same
          protein sequence, except now both the ambiguous character
          protection is off (false) and the unambiguous character
          protection status is also off.  Now, the text may be added to
          the protein sequence without problems.</li>
 </ol>
 <p><b><i>Please note that ON/OFF are interchangeable in this document
 with the terms TRUE/FALSE, where ON is TRUE, and OFF is FALSE.</i></b>
 </p>

 <p><b><i>Please also note that only DNA, RNA and PROTEIN sequences have
  protection settings, thus any non DNA, RNA, or PROTEIN sequence type
  tested will always return false, regardless of the text.  An example
  of a non-RNA/DNA/PROTEIN sequence is Seq.Type.TEXT.</i></b></p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>type</code> - the type of the sequence to test.</dd>
<dd><code>protect_align</code> - the protection status of alignment characters.</dd>
<dd><code>protect_ambig</code> - the protection status of ambiguous characters.</dd>
<dd><code>protect_uambig</code> - the protection status of unambiguous characters.</dd>
<dd><code>test</code> - the sequence text to test.</dd>
<dd><code>start</code> - the index within 'text' to start the test.</dd>
<dd><code>end</code> - the index within 'text' to end the test.</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>true if the text violates the protection settings of the
         sequence, and hence the text should not be inserted or deleted.</dd>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../../org/biolegato/sequence/data/Seq.html#protect_align"><code>Seq.protect_align</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#protect_ambig"><code>Seq.protect_ambig</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#protect_unambig"><code>Seq.protect_unambig</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.Type.html#DNA"><code>Seq.Type.DNA</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.Type.html#RNA"><code>Seq.Type.RNA</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.Type.html#PROTEIN"><code>Seq.Type.PROTEIN</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.Type.html#MASK"><code>Seq.Type.MASK</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.Type.html#TEXT"><code>Seq.Type.TEXT</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Dataset.html#insert-int-int-char:A-int-int-boolean-"><code>insert(int, int, char[], int, int, boolean)</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Dataset.html#delete-int-int-int-boolean-"><code>delete(int, int, int, boolean)</code></a></dd>
</dl>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<!-- ========= END OF CLASS DATA ========= -->
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/Dataset.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../index-all.html">Index</a></li>
<li><a href="../../../../help-doc.html">Help</a></li>
</ul>
<div class="aboutLanguage">BioLegato sequence canvas support v1.1.0 API</div>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../../../../org/biolegato/sequence/data/DataFormat.html" title="class in org.biolegato.sequence.data"><span class="typeNameLink">Prev&nbsp;Class</span></a></li>
<li><a href="../../../../org/biolegato/sequence/data/FastAFile.html" title="class in org.biolegato.sequence.data"><span class="typeNameLink">Next&nbsp;Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../../../../index.html?org/biolegato/sequence/data/Dataset.html" target="_top">Frames</a></li>
<li><a href="Dataset.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#fields.inherited.from.class.javax.swing.AbstractListModel">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
<p class="legalCopy"><small><i>Copyright &#169; 2008-2022 University of Manitoba.</i></small></p>
</body>
</html>
