<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (1.8.0_192) on Thu Aug 04 13:00:37 GMT-06:00 2022 -->
<title>Seq</title>
<meta name="date" content="2022-08-04">
<link rel="stylesheet" type="text/css" href="../../../../stylesheet.css" title="Style">
<script type="text/javascript" src="../../../../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="Seq";
        }
    }
    catch(err) {
    }
//-->
var methods = {"i0":10,"i1":9,"i2":10};
var tabs = {65535:["t0","All Methods"],1:["t1","Static Methods"],2:["t2","Instance Methods"],8:["t4","Concrete Methods"]};
var altColor = "altColor";
var rowColor = "rowColor";
var tableTab = "tableTab";
var activeTableTab = "activeTableTab";
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/Seq.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../index-all.html">Index</a></li>
<li><a href="../../../../help-doc.html">Help</a></li>
</ul>
<div class="aboutLanguage">BioLegato sequence canvas support v1.1.0 API</div>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../../../../org/biolegato/sequence/data/GenBankFile2008.html" title="class in org.biolegato.sequence.data"><span class="typeNameLink">Prev&nbsp;Class</span></a></li>
<li><a href="../../../../org/biolegato/sequence/data/Seq.Direction.html" title="enum in org.biolegato.sequence.data"><span class="typeNameLink">Next&nbsp;Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../../../../index.html?org/biolegato/sequence/data/Seq.html" target="_top">Frames</a></li>
<li><a href="Seq.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li><a href="#nested.class.summary">Nested</a>&nbsp;|&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="subTitle">org.biolegato.sequence.data</div>
<h2 title="Class Seq" class="title">Class Seq</h2>
</div>
<div class="contentContainer">
<ul class="inheritance">
<li>java.lang.Object</li>
<li>
<ul class="inheritance">
<li>org.biolegato.sequence.data.Seq</li>
</ul>
</li>
</ul>
<div class="description">
<ul class="blockList">
<li class="blockList">
<dl>
<dt>All Implemented Interfaces:</dt>
<dd>java.io.Serializable, java.lang.Cloneable</dd>
</dl>
<hr>
<br>
<pre>public final class <span class="typeNameLabel">Seq</span>
extends java.lang.Object
implements java.lang.Cloneable, java.io.Serializable</pre>
<div class="block"><p>This class is used to contain all sequence related functions.</p>

 <p><dl><dt>The main functionality of this class is to wrap the following
  parameters:</dt>
      <dd>name, type, topology, direction, strandedness, description,
      groupID, sequence, original, mask, protect_align, protect_ambig,
      and protect_unambig</dd></dl>
 </p>

 <p>Please see a description of each of these fields before making any major
 edits to this class.  Additionally, please note that the toString method
 should not be edited lightly (because it is relied upon by other classes).
 </p>

 <p><u>WHY PACKAGE ACCESS FOR MANY VARIABLES?</u></p>
 <!--  -------------------------------------- -->
 <p>Because the access is restricted to only the package level, there should
 not be any major issues with regard to code bugs.</p>
 <ul>
      <li>http://java.dzone.com/articles/getter-setter-use-or-not-use-0</li>
      <li>http://www.artima.com/intv/sway2.html</li>
 </ul>

 <p>In addition, because these variables do not have to be validated
 (they can be any value, including null), and the variables are only
 changed by classes and methods within the package there is no reason for
 accessor methods because it would only cause a slow-down
 (i.e. additional overhead).  In addition, the variables are only given
 package access to minimize the amount of code that can change them.</p>

 <p>Please note that any variable defined in the scope of a class without the
 word 'public' or 'private' preceding it, is defined as package scope.
 Package scope means that only classes and methods in the same package can
 alter or read the value directly.</p>

 <p><u>WHY TRANSIENT?</u></p>
 <!--  -------------- -->
 Certain parts of the object, such as the protection status and groupID,
 are not necessary to store when serializing this object.  To indicate
 this to Java, we add the word 'transient' before each variable
 declaration.</div>
<dl>
<dt><span class="simpleTagLabel">Author:</span></dt>
<dd>Graham Alvare, Brian Fristensky</dd>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../../serialized-form.html#org.biolegato.sequence.data.Seq">Serialized Form</a></dd>
</dl>
</li>
</ul>
</div>
<div class="summary">
<ul class="blockList">
<li class="blockList">
<!-- ======== NESTED CLASS SUMMARY ======== -->
<ul class="blockList">
<li class="blockList"><a name="nested.class.summary">
<!--   -->
</a>
<h3>Nested Class Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Nested Class Summary table, listing nested classes, and an explanation">
<caption><span>Nested Classes</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../org/biolegato/sequence/data/Seq.Direction.html" title="enum in org.biolegato.sequence.data">Seq.Direction</a></span></code>
<div class="block">Used for typing/storing sequence direction.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../org/biolegato/sequence/data/Seq.Strandedness.html" title="enum in org.biolegato.sequence.data">Seq.Strandedness</a></span></code>
<div class="block">Used for typing/storing sequence strandedness.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../org/biolegato/sequence/data/Seq.Topology.html" title="enum in org.biolegato.sequence.data">Seq.Topology</a></span></code>
<div class="block">Used for typing/storing sequence topology.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../org/biolegato/sequence/data/Seq.Type.html" title="enum in org.biolegato.sequence.data">Seq.Type</a></span></code>
<div class="block">This enum is used for typing/storing sequence types.</div>
</td>
</tr>
</table>
</li>
</ul>
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor.summary">
<!--   -->
</a>
<h3>Constructor Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Constructor Summary table, listing constructors, and an explanation">
<caption><span>Constructors</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../org/biolegato/sequence/data/Seq.html#Seq--">Seq</a></span>()</code>
<div class="block">Constructs new instances of a sequence object.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../org/biolegato/sequence/data/Seq.html#Seq-org.biolegato.sequence.data.Seq.Type-java.lang.String-java.lang.StringBuffer-">Seq</a></span>(<a href="../../../../org/biolegato/sequence/data/Seq.Type.html" title="enum in org.biolegato.sequence.data">Seq.Type</a>&nbsp;type,
   java.lang.String&nbsp;name,
   java.lang.StringBuffer&nbsp;sequence)</code>
<div class="block">Constructs new instances of a sequence object.</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../org/biolegato/sequence/data/Seq.html#Seq-org.biolegato.sequence.data.Seq.Type-java.lang.String-java.lang.StringBuffer-org.biolegato.sequence.data.Seq.Direction-org.biolegato.sequence.data.Seq.Topology-org.biolegato.sequence.data.Seq.Strandedness-">Seq</a></span>(<a href="../../../../org/biolegato/sequence/data/Seq.Type.html" title="enum in org.biolegato.sequence.data">Seq.Type</a>&nbsp;type,
   java.lang.String&nbsp;name,
   java.lang.StringBuffer&nbsp;sequence,
   <a href="../../../../org/biolegato/sequence/data/Seq.Direction.html" title="enum in org.biolegato.sequence.data">Seq.Direction</a>&nbsp;direction,
   <a href="../../../../org/biolegato/sequence/data/Seq.Topology.html" title="enum in org.biolegato.sequence.data">Seq.Topology</a>&nbsp;topology,
   <a href="../../../../org/biolegato/sequence/data/Seq.Strandedness.html" title="enum in org.biolegato.sequence.data">Seq.Strandedness</a>&nbsp;strandedness)</code>
<div class="block">Constructs new instances of a sequence object.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../org/biolegato/sequence/data/Seq.html#Seq-org.biolegato.sequence.data.Seq.Type-java.lang.String-java.lang.StringBuffer-org.biolegato.sequence.data.Seq.Direction-org.biolegato.sequence.data.Seq.Topology-org.biolegato.sequence.data.Seq.Strandedness-java.lang.CharSequence-">Seq</a></span>(<a href="../../../../org/biolegato/sequence/data/Seq.Type.html" title="enum in org.biolegato.sequence.data">Seq.Type</a>&nbsp;type,
   java.lang.String&nbsp;name,
   java.lang.StringBuffer&nbsp;sequence,
   <a href="../../../../org/biolegato/sequence/data/Seq.Direction.html" title="enum in org.biolegato.sequence.data">Seq.Direction</a>&nbsp;direction,
   <a href="../../../../org/biolegato/sequence/data/Seq.Topology.html" title="enum in org.biolegato.sequence.data">Seq.Topology</a>&nbsp;topology,
   <a href="../../../../org/biolegato/sequence/data/Seq.Strandedness.html" title="enum in org.biolegato.sequence.data">Seq.Strandedness</a>&nbsp;strandedness,
   java.lang.CharSequence&nbsp;original)</code>
<div class="block">Constructs new instances of a sequence object.</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../org/biolegato/sequence/data/Seq.html#Seq-org.biolegato.sequence.data.Seq.Type-java.lang.String-java.lang.StringBuffer-org.biolegato.sequence.data.Seq.Direction-org.biolegato.sequence.data.Seq.Topology-org.biolegato.sequence.data.Seq.Strandedness-int-java.lang.String-">Seq</a></span>(<a href="../../../../org/biolegato/sequence/data/Seq.Type.html" title="enum in org.biolegato.sequence.data">Seq.Type</a>&nbsp;type,
   java.lang.String&nbsp;name,
   java.lang.StringBuffer&nbsp;sequence,
   <a href="../../../../org/biolegato/sequence/data/Seq.Direction.html" title="enum in org.biolegato.sequence.data">Seq.Direction</a>&nbsp;direction,
   <a href="../../../../org/biolegato/sequence/data/Seq.Topology.html" title="enum in org.biolegato.sequence.data">Seq.Topology</a>&nbsp;topology,
   <a href="../../../../org/biolegato/sequence/data/Seq.Strandedness.html" title="enum in org.biolegato.sequence.data">Seq.Strandedness</a>&nbsp;strandedness,
   int&nbsp;groupID,
   java.lang.String&nbsp;description)</code>
<div class="block">Constructs new instances of a sequence object.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../org/biolegato/sequence/data/Seq.html#Seq-org.biolegato.sequence.data.Seq.Type-java.lang.String-java.lang.StringBuffer-java.lang.String-">Seq</a></span>(<a href="../../../../org/biolegato/sequence/data/Seq.Type.html" title="enum in org.biolegato.sequence.data">Seq.Type</a>&nbsp;type,
   java.lang.String&nbsp;name,
   java.lang.StringBuffer&nbsp;sequence,
   java.lang.String&nbsp;description)</code>
<div class="block">Constructs new instances of a sequence object.</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../org/biolegato/sequence/data/Seq.html#Seq-org.biolegato.sequence.data.Seq-">Seq</a></span>(<a href="../../../../org/biolegato/sequence/data/Seq.html" title="class in org.biolegato.sequence.data">Seq</a>&nbsp;data)</code>
<div class="block">Constructs new instance of a sequence object by copying data from
 another, pre-existing sequence object.</div>
</td>
</tr>
</table>
</li>
</ul>
<!-- ========== METHOD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="method.summary">
<!--   -->
</a>
<h3>Method Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Method Summary table, listing methods, and an explanation">
<caption><span id="t0" class="activeTableTab"><span>All Methods</span><span class="tabEnd">&nbsp;</span></span><span id="t1" class="tableTab"><span><a href="javascript:show(1);">Static Methods</a></span><span class="tabEnd">&nbsp;</span></span><span id="t2" class="tableTab"><span><a href="javascript:show(2);">Instance Methods</a></span><span class="tabEnd">&nbsp;</span></span><span id="t4" class="tableTab"><span><a href="javascript:show(8);">Concrete Methods</a></span><span class="tabEnd">&nbsp;</span></span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tr id="i0" class="altColor">
<td class="colFirst"><code>java.lang.Object</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../org/biolegato/sequence/data/Seq.html#clone--">clone</a></span>()</code>
<div class="block">Clones the current sequence object.</div>
</td>
</tr>
<tr id="i1" class="rowColor">
<td class="colFirst"><code>static <a href="../../../../org/biolegato/sequence/data/Seq.Type.html" title="enum in org.biolegato.sequence.data">Seq.Type</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../org/biolegato/sequence/data/Seq.html#detectType-java.lang.StringBuffer-">detectType</a></span>(java.lang.StringBuffer&nbsp;data)</code>
<div class="block">This function detects the type of a sequence (RNA, DNA, or protein).</div>
</td>
</tr>
<tr id="i2" class="altColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../org/biolegato/sequence/data/Seq.html#toString--">toString</a></span>()</code>
<div class="block">Creates a string representation of the Seq and its fields
 This representation is limited to '[GROUP#|_] NAME' because the toString
 method will be called when the sequence is displayed in any JList.</div>
</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.java.lang.Object">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;java.lang.Object</h3>
<code>equals, finalize, getClass, hashCode, notify, notifyAll, wait, wait, wait</code></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="details">
<ul class="blockList">
<li class="blockList">
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor.detail">
<!--   -->
</a>
<h3>Constructor Detail</h3>
<a name="Seq--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>Seq</h4>
<pre>public&nbsp;Seq()</pre>
<div class="block"><p>Constructs new instances of a sequence object.  The names of the
 constructor parameters passed correspond directly with the class fields
 of the same names (i.e. the parameter 'name' corresponds directly to the
 sequence object field 'name').</p>

 <p>In this case, the sequence object will be a nameless empty DNA
 sequence.</p></div>
</li>
</ul>
<a name="Seq-org.biolegato.sequence.data.Seq.Type-java.lang.String-java.lang.StringBuffer-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>Seq</h4>
<pre>public&nbsp;Seq(<a href="../../../../org/biolegato/sequence/data/Seq.Type.html" title="enum in org.biolegato.sequence.data">Seq.Type</a>&nbsp;type,
           java.lang.String&nbsp;name,
           java.lang.StringBuffer&nbsp;sequence)</pre>
<div class="block"><p>Constructs new instances of a sequence object.  The names of the
 constructor parameters passed correspond directly with the class fields
 of the same names (i.e. the parameter 'name' corresponds directly to the
 sequence object field 'name').</p>

 <p>The only thing really noteworthy is that if the sequence provided is
 empty, the character protection settings will all be set to false.  This
 functionality is handled by this specific constructor method (hence why
 all other constructor methods, except the clone/copy Seq constructor
 method, interface with this constructor).</p>

 <p>This constructor is called directly by BioLegato's GDE flat-file
 parser, and also unifies all of the other Seq object constructors (except
 the copy/clone Seq constructor).</p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>type</code> - the type of data to store in the sequence object.</dd>
<dd><code>name</code> - the name of the sequence object.</dd>
<dd><code>sequence</code> - the text sequence to store in the sequence object.</dd>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../../org/biolegato/sequence/data/Seq.html#type"><code>type</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#name"><code>name</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#sequence"><code>sequence</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#Seq--"><code>Seq()</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#Seq-org.biolegato.sequence.data.Seq.Type-java.lang.String-java.lang.StringBuffer-java.lang.String-"><code>Seq(org.biolegato.sequence.data.Seq.Type, java.lang.String, java.lang.StringBuffer, java.lang.String)</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#Seq-org.biolegato.sequence.data.Seq.Type-java.lang.String-java.lang.StringBuffer-org.biolegato.sequence.data.Seq.Direction-org.biolegato.sequence.data.Seq.Topology-org.biolegato.sequence.data.Seq.Strandedness-"><code>Seq(org.biolegato.sequence.data.Seq.Type, java.lang.String, java.lang.StringBuffer, org.biolegato.sequence.data.Seq.Direction, org.biolegato.sequence.data.Seq.Topology, org.biolegato.sequence.data.Seq.Strandedness)</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#Seq-org.biolegato.sequence.data.Seq.Type-java.lang.String-java.lang.StringBuffer-org.biolegato.sequence.data.Seq.Direction-org.biolegato.sequence.data.Seq.Topology-org.biolegato.sequence.data.Seq.Strandedness-java.lang.CharSequence-"><code>Seq(org.biolegato.sequence.data.Seq.Type, java.lang.String, java.lang.StringBuffer, org.biolegato.sequence.data.Seq.Direction, org.biolegato.sequence.data.Seq.Topology, org.biolegato.sequence.data.Seq.Strandedness, java.lang.CharSequence)</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#Seq-org.biolegato.sequence.data.Seq.Type-java.lang.String-java.lang.StringBuffer-org.biolegato.sequence.data.Seq.Direction-org.biolegato.sequence.data.Seq.Topology-org.biolegato.sequence.data.Seq.Strandedness-int-java.lang.String-"><code>Seq(org.biolegato.sequence.data.Seq.Type, java.lang.String, java.lang.StringBuffer, org.biolegato.sequence.data.Seq.Direction, org.biolegato.sequence.data.Seq.Topology, org.biolegato.sequence.data.Seq.Strandedness, int, java.lang.String)</code></a></dd>
</dl>
</li>
</ul>
<a name="Seq-org.biolegato.sequence.data.Seq.Type-java.lang.String-java.lang.StringBuffer-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>Seq</h4>
<pre>public&nbsp;Seq(<a href="../../../../org/biolegato/sequence/data/Seq.Type.html" title="enum in org.biolegato.sequence.data">Seq.Type</a>&nbsp;type,
           java.lang.String&nbsp;name,
           java.lang.StringBuffer&nbsp;sequence,
           java.lang.String&nbsp;description)</pre>
<div class="block"><p>Constructs new instances of a sequence object.  The names of the
 constructor parameters passed correspond directly with the class fields
 of the same names (i.e. the parameter 'name' corresponds directly to the
 sequence object field 'name').</p>

 <p>The only thing really noteworthy is that if the sequence provided is
 empty, the character protection settings will all be set to false.</p>

 <p>This constructor is called by BioLegato's FastA file parser.</p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>type</code> - the type of data to store in the sequence object.</dd>
<dd><code>name</code> - the name of the sequence object.</dd>
<dd><code>sequence</code> - the text sequence to store in the sequence object.</dd>
<dd><code>description</code> - the description of the sequence object.</dd>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../../org/biolegato/sequence/data/Seq.html#type"><code>type</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#name"><code>name</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#sequence"><code>sequence</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#description"><code>description</code></a>, 
<a href="../../../../org/biolegato/sequence/data/FastAFile.html" title="class in org.biolegato.sequence.data"><code>FastAFile</code></a></dd>
</dl>
</li>
</ul>
<a name="Seq-org.biolegato.sequence.data.Seq.Type-java.lang.String-java.lang.StringBuffer-org.biolegato.sequence.data.Seq.Direction-org.biolegato.sequence.data.Seq.Topology-org.biolegato.sequence.data.Seq.Strandedness-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>Seq</h4>
<pre>public&nbsp;Seq(<a href="../../../../org/biolegato/sequence/data/Seq.Type.html" title="enum in org.biolegato.sequence.data">Seq.Type</a>&nbsp;type,
           java.lang.String&nbsp;name,
           java.lang.StringBuffer&nbsp;sequence,
           <a href="../../../../org/biolegato/sequence/data/Seq.Direction.html" title="enum in org.biolegato.sequence.data">Seq.Direction</a>&nbsp;direction,
           <a href="../../../../org/biolegato/sequence/data/Seq.Topology.html" title="enum in org.biolegato.sequence.data">Seq.Topology</a>&nbsp;topology,
           <a href="../../../../org/biolegato/sequence/data/Seq.Strandedness.html" title="enum in org.biolegato.sequence.data">Seq.Strandedness</a>&nbsp;strandedness)</pre>
<div class="block"><p>Constructs new instances of a sequence object.  The names of the
 constructor parameters passed correspond directly with the class fields
 of the same names (i.e. the parameter 'name' corresponds directly to the
 sequence object field 'name').</p>

 <p>The only thing really noteworthy is that if the sequence provided is
 empty, the character protection settings will all be set to false.</p>

 <p>This method unifies the Seq constructors called by the GenBANK and
 GDE parsers.</p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>type</code> - the type of data to store in the sequence object.</dd>
<dd><code>name</code> - the name of the sequence object.</dd>
<dd><code>sequence</code> - the text sequence to store in the sequence object.</dd>
<dd><code>direction</code> - the direction of the sequence (if nucleotide).</dd>
<dd><code>topology</code> - the topology of the sequence (if nucleotide).</dd>
<dd><code>strandedness</code> - the strandedness of the sequence (if nucleotide).</dd>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../../org/biolegato/sequence/data/Seq.html#type"><code>type</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#name"><code>name</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#sequence"><code>sequence</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#direction"><code>direction</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#topology"><code>topology</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#strandedness"><code>strandedness</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#Seq-org.biolegato.sequence.data.Seq.Type-java.lang.String-java.lang.StringBuffer-org.biolegato.sequence.data.Seq.Direction-org.biolegato.sequence.data.Seq.Topology-org.biolegato.sequence.data.Seq.Strandedness-java.lang.CharSequence-"><code>Seq(org.biolegato.sequence.data.Seq.Type, java.lang.String, java.lang.StringBuffer, org.biolegato.sequence.data.Seq.Direction, org.biolegato.sequence.data.Seq.Topology, org.biolegato.sequence.data.Seq.Strandedness, java.lang.CharSequence)</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#Seq-org.biolegato.sequence.data.Seq.Type-java.lang.String-java.lang.StringBuffer-org.biolegato.sequence.data.Seq.Direction-org.biolegato.sequence.data.Seq.Topology-org.biolegato.sequence.data.Seq.Strandedness-int-java.lang.String-"><code>Seq(org.biolegato.sequence.data.Seq.Type, java.lang.String, java.lang.StringBuffer, org.biolegato.sequence.data.Seq.Direction, org.biolegato.sequence.data.Seq.Topology, org.biolegato.sequence.data.Seq.Strandedness, int, java.lang.String)</code></a></dd>
</dl>
</li>
</ul>
<a name="Seq-org.biolegato.sequence.data.Seq.Type-java.lang.String-java.lang.StringBuffer-org.biolegato.sequence.data.Seq.Direction-org.biolegato.sequence.data.Seq.Topology-org.biolegato.sequence.data.Seq.Strandedness-java.lang.CharSequence-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>Seq</h4>
<pre>public&nbsp;Seq(<a href="../../../../org/biolegato/sequence/data/Seq.Type.html" title="enum in org.biolegato.sequence.data">Seq.Type</a>&nbsp;type,
           java.lang.String&nbsp;name,
           java.lang.StringBuffer&nbsp;sequence,
           <a href="../../../../org/biolegato/sequence/data/Seq.Direction.html" title="enum in org.biolegato.sequence.data">Seq.Direction</a>&nbsp;direction,
           <a href="../../../../org/biolegato/sequence/data/Seq.Topology.html" title="enum in org.biolegato.sequence.data">Seq.Topology</a>&nbsp;topology,
           <a href="../../../../org/biolegato/sequence/data/Seq.Strandedness.html" title="enum in org.biolegato.sequence.data">Seq.Strandedness</a>&nbsp;strandedness,
           java.lang.CharSequence&nbsp;original)</pre>
<div class="block"><p>Constructs new instances of a sequence object.  The names of the
 constructor parameters passed correspond directly with the class fields
 of the same names (i.e. the parameter 'name' corresponds directly to the
 sequence object field 'name').</p>

 <p>The only thing really noteworthy is that if the sequence provided is
 empty, the character protection settings will all be set to false.</p>

 <p>This constructor is called by BioLegato's GenBANK parser.</p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>type</code> - the type of data to store in the sequence object.</dd>
<dd><code>name</code> - the name of the sequence object.</dd>
<dd><code>sequence</code> - the text sequence to store in the sequence object.</dd>
<dd><code>direction</code> - the direction of the sequence (if nucleotide).</dd>
<dd><code>topology</code> - the topology of the sequence (if nucleotide).</dd>
<dd><code>strandedness</code> - the strandedness of the sequence (if nucleotide).</dd>
<dd><code>original</code> - the original GenBANK header for the sequence data.</dd>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../../org/biolegato/sequence/data/Seq.html#type"><code>type</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#name"><code>name</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#sequence"><code>sequence</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#direction"><code>direction</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#topology"><code>topology</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#strandedness"><code>strandedness</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#original"><code>original</code></a>, 
<a href="../../../../org/biolegato/sequence/data/GenBankFile2008.html" title="class in org.biolegato.sequence.data"><code>GenBankFile2008</code></a></dd>
</dl>
</li>
</ul>
<a name="Seq-org.biolegato.sequence.data.Seq.Type-java.lang.String-java.lang.StringBuffer-org.biolegato.sequence.data.Seq.Direction-org.biolegato.sequence.data.Seq.Topology-org.biolegato.sequence.data.Seq.Strandedness-int-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>Seq</h4>
<pre>public&nbsp;Seq(<a href="../../../../org/biolegato/sequence/data/Seq.Type.html" title="enum in org.biolegato.sequence.data">Seq.Type</a>&nbsp;type,
           java.lang.String&nbsp;name,
           java.lang.StringBuffer&nbsp;sequence,
           <a href="../../../../org/biolegato/sequence/data/Seq.Direction.html" title="enum in org.biolegato.sequence.data">Seq.Direction</a>&nbsp;direction,
           <a href="../../../../org/biolegato/sequence/data/Seq.Topology.html" title="enum in org.biolegato.sequence.data">Seq.Topology</a>&nbsp;topology,
           <a href="../../../../org/biolegato/sequence/data/Seq.Strandedness.html" title="enum in org.biolegato.sequence.data">Seq.Strandedness</a>&nbsp;strandedness,
           int&nbsp;groupID,
           java.lang.String&nbsp;description)</pre>
<div class="block"><p>Constructs new instances of a sequence object.  The names of the
 constructor parameters passed correspond directly with the class fields
 of the same names (i.e. the parameter 'name' corresponds directly to the
 sequence object field 'name').</p>

 <p>The only thing really noteworthy is that if the sequence provided is
 empty, the character protection settings will all be set to false.</p>

 <p>This constructor is called by BioLegato's GDE parser.</p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>type</code> - the type of data to store in the sequence object.</dd>
<dd><code>name</code> - the name of the sequence object.</dd>
<dd><code>sequence</code> - the text sequence to store in the sequence object.</dd>
<dd><code>direction</code> - the direction of the sequence (if nucleotide).</dd>
<dd><code>topology</code> - the topology of the sequence (if nucleotide).</dd>
<dd><code>strandedness</code> - the strandedness of the sequence (if nucleotide).</dd>
<dd><code>groupID</code> - the group ID for the sequence object.</dd>
<dd><code>description</code> - the description of the sequence object.</dd>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../../org/biolegato/sequence/data/Seq.html#type"><code>type</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#name"><code>name</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#sequence"><code>sequence</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#direction"><code>direction</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#topology"><code>topology</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#strandedness"><code>strandedness</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#groupID"><code>groupID</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#description"><code>description</code></a>, 
<a href="../../../../org/biolegato/sequence/data/GDEFile.html" title="class in org.biolegato.sequence.data"><code>GDEFile</code></a></dd>
</dl>
</li>
</ul>
<a name="Seq-org.biolegato.sequence.data.Seq-">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>Seq</h4>
<pre>public&nbsp;Seq(<a href="../../../../org/biolegato/sequence/data/Seq.html" title="class in org.biolegato.sequence.data">Seq</a>&nbsp;data)</pre>
<div class="block">Constructs new instance of a sequence object by copying data from
 another, pre-existing sequence object.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>data</code> - the sequence to copy data from.</dd>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../../org/biolegato/sequence/data/Seq.html#clone--"><code>clone()</code></a></dd>
</dl>
</li>
</ul>
</li>
</ul>
<!-- ============ METHOD DETAIL ========== -->
<ul class="blockList">
<li class="blockList"><a name="method.detail">
<!--   -->
</a>
<h3>Method Detail</h3>
<a name="toString--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>toString</h4>
<pre>public&nbsp;java.lang.String&nbsp;toString()</pre>
<div class="block"><p>Creates a string representation of the Seq and its fields
 This representation is limited to '[GROUP#|_] NAME' because the toString
 method will be called when the sequence is displayed in any JList.</p>

 <p>A couple of examples are as follows:</p>
 <pre>
      1 HEXOKINASE
      1 PYRUVATE KINASE
      1 ENOLASE
      2 ACONITASE
      _ OTC</pre>

 <p>
 Where the glycolysis enzymes (HEXOKINASE, PYRUVATE KINASE, and ENOLASE)
 are in group 1, the citric acid enzyme (ACONITASE) is in group 2, and
 the urea cycle enzyme (OTC -- ornithine transcarbamylase) is not grouped.
 </p>

 <p>Reliability and performance are the reasons why the string format
 limitation of '[GROUP#|_] NAME' was imposed:</p>
 <ol>
  <li>Reliability: the code will provide a simpler implementation for
      displaying sequences in a list.  Instead of writing a custom class
      to display a list of sequences or using other possible complicated
      implementations (e.g. writing multiple external methods to "wrap"
      the sequence), the toString method and Java's intrinsic JList
      provide a simple streamlined way to display a list of sequences.
      Because the code is simpler, it should be more reliable.</li>
  <li>Performance: because the code is simpler and can be accessed
      directly by the Java API (no "wrapping" necessary), the amount of
      Java code required to display the sequence list is minimal.
      Additionally, JList's implementation is likely (but not necessarily)
      optimized C/C++ code, which should (theoretically) be much faster
      than any practical custom JComponent display list code.</li>
  <li>Readability: smaller code is almost always more readable (as long as
      it is documented properly).</li>
 </ol></div>
<dl>
<dt><span class="overrideSpecifyLabel">Overrides:</span></dt>
<dd><code>toString</code>&nbsp;in class&nbsp;<code>java.lang.Object</code></dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>a string containing the name and group number of the sequence.
         The format is as follows: '[GROUP#|_] NAME'.</dd>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../../org/biolegato/sequence/data/Seq.html#groupID"><code>groupID</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Seq.html#name"><code>name</code></a>, 
<a href="../../../../org/biolegato/sequence/data/Dataset.html#getElementAt-int-"><code>Dataset.getElementAt(int)</code></a></dd>
</dl>
</li>
</ul>
<a name="clone--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>clone</h4>
<pre>public&nbsp;java.lang.Object&nbsp;clone()</pre>
<div class="block">Clones the current sequence object.</div>
<dl>
<dt><span class="overrideSpecifyLabel">Overrides:</span></dt>
<dd><code>clone</code>&nbsp;in class&nbsp;<code>java.lang.Object</code></dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>a cloned copy of this sequence object.</dd>
</dl>
</li>
</ul>
<a name="detectType-java.lang.StringBuffer-">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>detectType</h4>
<pre>public static&nbsp;<a href="../../../../org/biolegato/sequence/data/Seq.Type.html" title="enum in org.biolegato.sequence.data">Seq.Type</a>&nbsp;detectType(java.lang.StringBuffer&nbsp;data)</pre>
<div class="block"><p>This function detects the type of a sequence (RNA, DNA, or protein).
 </p>

 <p>Please note that this function's implementation is very simple and may
 not be accurate in all cases.  This function reads sequence data and
 looks for the characters F, E, J, L, O, Q, X, Z, or U.  If the characters
 F, E, J, L, O, Q, X, or Z are encountered, the sequence is determined to
 be an amino acid/protein sequence; however, if the character U is found,
 the sequence is determined to be RNA.  If both the character U and any
 character from the protein set (F, E, J, L, O, Q, X, or Z) are found,
 then the first of those characters found will be used to determine the
 sequence type.  Note that if none of those characters (F, E, J, L, O, Q,
 X, Z, or U) are detected, then the sequence type will default to DNA.</p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>data</code> - the sequence object to detect the type for.</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the sequence type detected.</dd>
</dl>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<!-- ========= END OF CLASS DATA ========= -->
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/Seq.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../index-all.html">Index</a></li>
<li><a href="../../../../help-doc.html">Help</a></li>
</ul>
<div class="aboutLanguage">BioLegato sequence canvas support v1.1.0 API</div>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../../../../org/biolegato/sequence/data/GenBankFile2008.html" title="class in org.biolegato.sequence.data"><span class="typeNameLink">Prev&nbsp;Class</span></a></li>
<li><a href="../../../../org/biolegato/sequence/data/Seq.Direction.html" title="enum in org.biolegato.sequence.data"><span class="typeNameLink">Next&nbsp;Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../../../../index.html?org/biolegato/sequence/data/Seq.html" target="_top">Frames</a></li>
<li><a href="Seq.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li><a href="#nested.class.summary">Nested</a>&nbsp;|&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
<p class="legalCopy"><small><i>Copyright &#169; 2008-2022 University of Manitoba.</i></small></p>
</body>
</html>
